feat(product-assistant): product memory

skoob13 opened the pull request:
## Problem

LLMs are stateless, so our agents are dumb when questions require details. This PR gives them the ability to remember simple facts about a user's company or product in the core memory. The core memory is a text that the background agent continuously updates using two actions: appending a new fragment and replacing an existing fragment. The memory agent triggers on each new human message (surprisingly cost-effective).

Additionally, the PR includes another agent scraping the initial memory from the Internet. It searches in taxonomy for existing $host (preferable) or $app_namespace (app bundle id) and scrapes relevant sources using Perplexity to automatically fill the agent's initial core memory. However, this information is not always reliable, so we always confirm that we've scraped the memory correctly.

**Demos**

![2025-01-07 18 22 15](https://github.com/user-attachments/assets/aa3856a3-f07c-4e58-ba4a-c34d924f9b02)

![2025-01-07 18 26 35](https://github.com/user-attachments/assets/2e0cdcee-f00b-4b8f-8e24-2837f9ecf9c4)

## Changes

- Add onboarding for the agent.
- Scrape Perplexity for relevant memory about the user's company, product, business model, target audience, and features.
- Implement simple dynamic forms for confirmation/rejection of scraped info.
- Compress the scraped memory to a smaller text since Perplexity's output is **very** verbose.
- Implement the memory collector agent, which works as a parallel node on every new human message.
- Replace the product description with the core memory in the agent nodes.

## Does this work well for both Cloud and self-hosted?

No

## How did you test this code?

Manual testing, unit tests.

skoob13 committed:
feat: perplexity prompt

skoob13 committed:
chore: migrations

skoob13 committed:
feat: initial memory scraping

skoob13 committed:
feat: block onboarding when someone has started it

skoob13 committed:
feat: assistant onboarding

skoob13 committed:
feat: flag for resumed conversation

skoob13 committed:
fix: mypy

skoob13 committed:
chore: migration

skoob13 committed:
feat: stricter prompt for perplexity

skoob13 committed:
feat: simple dynamic forms

skoob13 committed:
feat: style form options

skoob13 committed:
feat: formatting core memory

skoob13 committed:
feat: core memory prompts

skoob13 committed:
feat: compress memory

skoob13 committed:
fix: update mobile prompts

skoob13 committed:
fix: disable streaming for compressor

skoob13 committed:
feat: memory collector nodes

skoob13 committed:
feat: append/replace memory

skoob13 committed:
fix: mypy

skoob13 committed:
feat: memory router

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
feat: re-use prompts

skoob13 committed:
fix: excessive detection

skoob13 committed:
test: eval tests

skoob13 committed:
fix: broader check for the planner

skoob13 committed:
fix: filter out summarizer messages

skoob13 committed:
fix: assistant tests

skoob13 committed:
test: core memory model

skoob13 committed:
Merge branch 'feat/core-agent-memory' of github.com:PostHog/posthog into feat/core-agent-memory

skoob13 committed:
test: onboarding node

skoob13 committed:
fix: allow optional state

skoob13 committed:
test: memory initializer

skoob13 committed:
test: memory initializer interrupt

skoob13 committed:
test: memory collector node

skoob13 committed:
test: memory collector tools

skoob13 committed:
test: abstract nodes

skoob13 committed:
fix: use product description from project for initial memory

skoob13 committed:
fix: rename tests

skoob13 committed:
Merge branch 'fix/taxonomy-planner-failover' into feat/core-agent-memory

skoob13 committed:
fix: set skipped status after rejecting scraped memory

skoob13 committed:
test: assistant tests

skoob13 committed:
chore: document code and move out messages to prompts

skoob13 committed:
fix: recompile dev requirements

skoob13 committed:
fix: mypy

skoob13 committed:
fix: hide form actions after submitting a response

skoob13 committed:
fix: missing messages from memory initializer

skoob13 committed:
Merge branch 'master' into feat/core-agent-memory

Twixes committed:
Merge branch 'master' into feat/core-agent-memory

Twixes commented on frontend/src/scenes/max/maxLogic.ts:
```diff
@@ -318,6 +318,20 @@ export const maxLogic = kea<maxLogicType>([
+                return false
+            },
+        ],
+        inputDisabled: [
```
Would be smoother UX if we turned this into `inputDisabledReason`, so that in the form pending case we could show something like "Please choose one of the proposed answers"

Twixes replied:
Ah, since this is on the text area, which doesn't have `disabledReason`, let's skip

Twixes commented on frontend/src/scenes/max/Thread.tsx:
```diff
@@ -100,8 +101,8 @@ function MessageGroup({ messages, isFinal: isGroupFinal, index: messageGroupInde
-                                rateable={messageIndex === messages.length - 1}
-                                retriable={messageIndex === messages.length - 1 && isGroupFinal}
+                                interactable={messageIndex === messages.length - 1}
+                                finalGroup={isGroupFinal}
```
Let's do `isFinalGroup={isFinal}` for a bit more consistency

Twixes committed:
Prevent memory being left in pending state

Twixes committed:
Tweak wording of the initialization flow

Twixes committed:
Tweak initialization prompts

Been testing initialization with demo data containing `dropbox.com`. The instructions around "sites like Crunchbase" were confusing the Sonar model, as it was often thinking we're looking at a product _like_ Crunchbase etc. This seems to improve the handling.

Twixes committed:
Merge branch 'master' into feat/core-agent-memory

Twixes committed:
Unify `isFinalGroup` naming

Twixes commented on ee/hogai/summarizer/nodes.py:
```diff
@@ -84,7 +84,7 @@ def run(self, state: AssistantState, config: RunnableConfig) -> PartialAssistant
             {
                 "query_kind": viz_message.answer.kind,
-                "product_description": self._team.project.product_description,
+                "product_description": self.product_core_memory,
```
Should this become `core_memory` too?

Twixes reviewed (approved):
Pushed a few tweaks which should make things work even better. Overall, beautiful stuff

Twixes committed:
Fix missing __init__.py in memory tests

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

Twixes committed:
Add one more __init__.py

Twixes committed:
Merge branch 'master' into feat/core-agent-memory

github-actions[bot] committed:
Update UI snapshots for `chromium` (1)

skoob13 committed:
fix: review comments

skoob13 committed:
Merge branch 'master' of github.com:PostHog/posthog into feat/core-agent-memory

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)

github-actions[bot] committed:
Update UI snapshots for `chromium` (2)
